### DISCUSSION TITLE
How to vmap stateful single-example models

### DISCUSSION THREAD

**Discussion by JamesAllingham** [google/flax#1248] [Q&A] [answered]
Hi all,

I'm a little stuck with how to define models which work on single examples and then `vmap` them across a batch, when that model has some modifiable state (e.g. BN layers). 

For a stateless single-example model, I can do something like this:

```
def update_step(apply_fun, x_batch, y_batch, optimizer):
  def batch_loss(params):
    def loss_fn(x, y):
      preds = apply_fun({'params': params}, x)
      return mse(y, preds)
    
    loss = jax.vmap(loss_fn)(x_batch, y_batch)
    return jnp.mean(loss, axis=0)

  loss, grads = jax.value_and_grad(batch_loss)(optimizer.target)
  optimizer = optimizer.apply_gradient(grads)
  return optimizer, loss

for _ in range(iters):
  optimizer, loss = update_step(model.apply, x_batch, y_batch, optimizer)
```

However, I am not sure how this should be adapted to handle stateful models. Unfortunately, [the guide for handling state](https://flax.readthedocs.io/en/latest/howtos/state_params.html) doesn't touch on `vmap`. 

Naively trying to combine the approach in the guide and my code above leads me to something like this:

```
def update_step(apply_fun, x_batch, y_batch, optimizer, state):
  def batch_loss(params):
    def loss_fn(x, y):
      preds, updated_state = apply_fun({'params': params, **state},
                                  x, mutable=list(state.keys()))
      return mse(y, preds), updated_state
    
    loss, updated_state = jax.vmap(loss_fn)(x_batch, y_batch)
    return jnp.mean(loss, axis=0), updated_state
    
  (loss, updated_state), grads = jax.value_and_grad(
      batch_loss, has_aux=True)(optimizer.target)
  optimizer = optimizer.apply_gradient(grads)
  return optimizer, updated_state, loss

for _ in range(iters):
  optimizer, state, loss = update_step(model.apply, x_batch, y_batch, optimizer, state)
```
But unfortunately, this doesn't work because `updated_state` is now duplicated for each example in the batch. A simple example to illustrate this kind of behaviour (with the caveat that I don't know how things are handled under to hood of Flax/Jax so maybe this isn't a very accurate reflection of reality):
```
>>> print(vmap(lambda x, state: (x + state["y"], flax.core.freeze({"y": state["y"] * 2.})), in_axes=(0, None))(jnp.arange(3.), {"y": 4.}))
(DeviceArray([4., 5., 6.], dtype=float32), FrozenDict({
    y: DeviceArray([8., 8., 8.], dtype=float32),
}))
```
This leads to strange training behaviour and memory issues and is clearly not what we want. So the next naive idea is to make a minor modification to the `vmap` call: 
```
>>> print(vmap(lambda x, state: (x + state["y"], flax.core.freeze({"y": state["y"] * 2.})), in_axes=(0, None), out_axes=(0, None))(jnp.arange(3.), {"y": 4.}))
(DeviceArray([4., 5., 6.], dtype=float32), FrozenDict({
    y: 8.0,
}))
```
BUT there are two problems. Firstly, if I try to do this in a real example (see my notebook below) it doesn't work:
```
ValueError: vmap has mapped output but out_axes is None
```
So I am guessing that my illustrative example hasn't captured reality very well. 

The second issue with this kind of fix is that it doesn't really make sense since the state will depend on the BN layer logic and on the whole batch of data...  and it isn't clear to me that `vmap` can handle this.

Another solution that crossed my mind was modifying the state returned by the `vmap`ed function. However, this doesn't seem like a good idea since a) it isn't clear how to aggregate the state or select a subset, and b) modifying a `FrozenDict` seems like a bad idea :)  

So I am now wondering if this is actually possible? Given that 'write models as “single-example” code and introduce batching automatically with `vmap`' is a desideratum mentioned in the Linen design docs, I would assume that it is possible and that I've missed something. Or maybe I am doing something very foolish... here is a notebook I was using to try to figure this all out: https://colab.research.google.com/drive/16WgbBrQPeYyDRODPZDyZnZ0eePN_eIbd?usp=sharing

If it is possible, perhaps an example would help here? 

(Apologies for the long wall of text! I could probably have written this more succinctly but I wrote this as I was trying to figure it all out!)

**Comment by jheek** [comment, marked as answer]
The problem you are running into is that BatchNorm isn't vectorizable strictly speaking. In order to share the batch statistics we have to sum over the batch dimmension but it has disappeared because you map over it. The way to solve this is by using `BatchNorm(axis_name="batch")` and use `vmap(loss_fn, out_axes=(0, None), axis_name="batch")`. The BatchNorm implementation will use `lax.pmean()` to average the statistics over the batch dimension so now they are in sync for each item in the batch.

**Reply by JamesAllingham** [reply]
Fantastic  – that works! And it is a super simple change. Thanks for the quick response. 

Could you elaborate on the meaning of "BatchNorm isn't vectorizable strictly speaking"? Specifically, I am wondering whether there are any gotchas that I should be aware of when using your solution?

**Reply by jheek** [reply]
Normally vectorization is defined as `vmap(f)(xs) == stack([f(x) for x in xs])` but for BatchNorm this does not hold.

You don't have to worry that it's wrong somehow implementation wise. But BatchNorm is odd in the sense that items in the batch communicate with each other over the batch dimension. Aside from BatchNorm this is never the case and many people have tried to make BatchNorm redundant for this reason (it turns out difficult to replace with something else though).

**Reply by JamesAllingham** [reply]
Great, thanks for the explanation. 

Haha yeah, I am looking forward to the day that I don't have to deal with BatchNorm! It always makes code messier, and causes a whole bunch of problems when trying to work with Bayesian neural nets!

**Reply by JamesAllingham** [reply]
By the way, I'd be happy to make a PR to add an example for how to deal with `vmap` + BN state to the [handling state guide](https://flax.readthedocs.io/en/latest/howtos/state_params.html)... would that be helpful?

**Reply by jheek** [reply]
>> and causes a whole bunch of problems when trying to work with Bayesian neural nets!

Well in that case maybe you should just use NFNet instead. You could even redefine their adaptive gradient clipping in a Bayesian way... I see a publication approaching ;)

>> By the way, I'd be happy to make a PR to add an example for how to deal with vmap + BN state to the handling state guide... would that be helpful?

sounds good.

**Comment by wztdream** [comment]
> The problem you are running into is that BatchNorm isn't vectorizable strictly speaking. In order to share the batch statistics we have to sum over the batch dimmension but it has disappeared because you map over it. The way to solve this is by using `BatchNorm(axis_name="batch")` and use `vmap(loss_fn, out_axes=(0, None), axis_name="batch")`. The BatchNorm implementation will use `lax.pmean()` to average the statistics over the batch dimension so now they are in sync for each item in the batch.

@jheek can you take a look at [this](https://github.com/google/flax/discussions/1675) similar issue about `flax lifted vmap` when deal with `BatchNorm`, it seems the method provided here not work.
